### ISSUE TITLE
Code fix to handle IGMP packet

### ISSUE THREAD

**Issue by GoogleCodeExporter** [CoreSecurity/impacket#30]
```
Steps need to properly handle IGMP packet:

==ImpactDecoder.py==
add at line 91:
        elif i.get_ip_p() == ImpactPacket.IGMP.protocol:
            self.igmp_decoder = IGMPDecoder()
            packet = self.igmp_decoder.decode(aBuffer[off:])

add at line 171:
class IGMPDecoder(Decoder):
    def __init__(self):
        pass

    def decode(self, aBuffer):
        ig = ImpactPacket.IGMP(aBuffer)
        off = ig.get_header_size()
        self.data_decoder = DataDecoder()
        packet = self.data_decoder.decode(aBuffer[off:])
        ig.contains(packet)
        return ig

==ImpactPacket.py==
change line 1670:
answer = tmp_dict.get(aType, 'UNKNOWN TYPE OR VERSION ')

change line 1680:
tmp_str += 'Group: ' + socket.inet_ntoa(struct.pack('!L',self.get_igmp_group()))

Hope I haven't missed anything!
```

Original issue reported on code.google.com by `andrewrw...@gmail.com` on 1 Jul 2013 at 3:26

**Comment by nixotic** [comment]
Hey, not sure whether anything ever happened with this, but I recently ran into the same issue.
I've implemented the above on 9.16 as best I could, diff below:

diff --git a/impacket/ImpactDecoder.py b/impacket/ImpactDecoder.py
index 11a1276..de81e03 100644
--- a/impacket/ImpactDecoder.py
+++ b/impacket/ImpactDecoder.py
@@ -138,6 +138,9 @@ class IPDecoder(Decoder):
         elif i.get_ip_p() == ImpactPacket.ICMP.protocol:
             self.icmp_decoder = ICMPDecoder()
             packet = self.icmp_decoder.decode(aBuffer[off:end])
+        elif i.get_ip_p() == ImpactPacket.IGMP.protocol:  # GSWAN 20160914
+            self.igmp_decoder = IGMPDecoder()  # GSWAN 20160914
+            packet = self.igmp_decoder.decode(aBuffer[off:end])  # GSWAN 20160914
         else:
             self.data_decoder = DataDecoder()
             packet = self.data_decoder.decode(aBuffer[off:end])
@@ -290,6 +293,18 @@ class TCPDecoder(Decoder):
         t.contains(packet)
         return t
 
+class IGMPDecoder(Decoder):  # GSWAN 20160914
+    def __init__(self):  # GSWAN 20160914
+        pass  # GSWAN 20160914
+    def decode(self, aBuffer):  # GSWAN 20160914
+        ig = ImpactPacket.IGMP(aBuffer)  # GSWAN 20160914
+        off = ig.get_header_size()  # GSWAN 20160914
+        self.data_decoder = DataDecoder()  # GSWAN 20160914
+        packet = self.data_decoder.decode(aBuffer[off:])  # GSWAN 20160914
+        ig.contains(packet)  # GSWAN 20160914
+        return ig  # GSWAN 20160914
+
+
 class IPDecoderForICMP(Decoder):
     """This class was added to parse the IP header of ICMP unreachables packets
     If you use the "standard" IPDecoder, it might crash (see bug #4870) ImpactPacket.py
diff --git a/impacket/version.py b/impacket/version.py
index a61468c..aab15d6 100644
--- a/impacket/version.py
+++ b/impacket/version.py
@@ -6,7 +6,7 @@
 #
 
 VER_MAJOR = "0"
-VER_MINOR = "9.16-dev"
+VER_MINOR = "9.17-dev"
 
 BANNER = "Impacket v%s.%s - Copyright 2002-2016 Core Security Technologies\n" % (VER_MAJOR,VER_MINOR)

**Comment by asolino** [comment]
Hey @nixotic thanks a lot for the patch!.. I just applied it. 

Closing now.
